torch/csrc/jit/codegen/cuda/transform_iter.cpp

```diff
@@ -440,6 +473,282 @@ int BestEffortReplay::findFirstMismatchedID(
+        FusionGuard::getCurFusion(),
+        std::vector<Val*>(
+            consumer->domain()->domain().begin(),
+            consumer->domain()->domain().end()));
+
+    auto isValInConsumer =
+        [&consumer_bcast_ids_not_in_producer](Val* input_val) {
+          if (input_val->isA<IterDomain>()) {
+            return consumer_bcast_ids_not_in_producer.find(
+                       input_val->as<IterDomain>()) !=
+                consumer_bcast_ids_not_in_producer.end();
+          } else {
+            return true;
+          }
+        };
+
+    for (auto expr : consumer_history) {
+      // If expr inputs are all in consumer_bcast_ids_not_in_producer, than so
+      // are all outputs
+      if (std::all_of(
+              expr->inputs().begin(), expr->inputs().end(), isValInConsumer)) {
+        // add all outputs to not being in producer
+        for (auto output_ids :
+             ir_utils::filterByType<IterDomain>(expr->outputs())) {
+          consumer_bcast_ids_not_in_producer.emplace(output_ids);
+        }
+      } else if (
+          expr->isA<Merge>() &&
+          std::any_of(
+              expr->inputs().begin(), expr->inputs().end(), isValInConsumer)) {
+        auto merge_expr = expr->as<Merge>();
+        // If
+        // - one of the inputs is made of id's in consumer that don't map to
+        // producer (bcast axes),
+        // - && the other input maps to an id in both consumer and producer
+        // - && this is a merge
+        //   for the sake of BestEffortReplay we can forward the input mapping
+        //   to both consumer and producer to the output of the expression
+        std::vector<IterDomain*> forwarded_ids;
+        std::vector<IterDomain*> compliment_ids;
+
+        for (auto input_id :
+             ir_utils::filterByType<IterDomain>(expr->inputs())) {
+          if (consumer_bcast_ids_not_in_producer.find(input_id) ==
+              consumer_bcast_ids_not_in_producer.end()) {
+            forwarded_ids.emplace_back(input_id);
+            forwarding_map.emplace(std::make_pair(input_id, merge_expr->out()));
+          } else {
+            compliment_ids.push_back(input_id);
+          }
+        }
+
+        // Set up compliment map
+        for (auto forwarded_id : forwarded_ids) {
+          compliment_map.emplace(std::make_pair(forwarded_id, compliment_ids));
+        }
+      }
+    }
+  }
+};
+
+} // namespace
+
+BestEffortReplay BestEffortReplay::replayCasP(
+    const TensorView* consumer,
+    const TensorView* producer,
+    int producer_compute_at_axis,
+    const RootDomainMap& root_map) {
+  if (producer_compute_at_axis < 0)
+    producer_compute_at_axis += (int)producer->nDims() + 1;
+
+  TORCH_INTERNAL_ASSERT(
+      producer_compute_at_axis >= 0 &&
+          (unsigned int)producer_compute_at_axis <= producer->nDims(),
+      "Invalid axis provided to BestEffortReplay::replayCasP.");
+
+  // Copied from TransformReplay::replayCasP
+  // TODO: Remove duplicate code
```

Comment:
Should remove this TODO. There is quite a bit of duplicate code, but it didn't seem like there was much that could be cleaned up as this function isn't only called from transform replay, so we can't remove logic from here.